SpatialGDK/Source/SpatialGDK/Private/Tests/RPCContainerTest.cpp

```diff
@@ -0,0 +1,232 @@
+	TestTrue("Queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRefReliable.Entity, ReliableType));
+
+    return true;
+}
+
+TEST(FRPCContainer, GIVEN_a_container_WHEN_two_values_of_different_type_have_been_added_and_processed_THEN_nothing_is_queued)
+{
+	UMockObject* TargetObject = NewObject<UMockObject>();
+	FUnrealObjectRef ObjecRef = GenerateObjectRef(TargetObject);
+	FPendingRPCParamsPtr ParamsUnreliable = CreateMockParameters(TargetObject, UnreliableType);
+	FPendingRPCParamsPtr ParamsReliable = CreateMockParameters(TargetObject, ReliableType);
+
+	FRPCContainer RPCs;
+
+	RPCs.QueueRPC(MoveTemp(ParamsUnreliable), UnreliableType);
+	RPCs.QueueRPC(MoveTemp(ParamsReliable), ReliableType);
+
+	FProcessRPCDelegate Delegate;
+	Delegate.BindUObject(TargetObject, &UMockObject::ProcessRPC);
+	RPCs.ProcessRPCs(Delegate);
+
+	TestFalse("No queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRef.Entity, UnreliableType));
+	TestFalse("No queued RPCs of such type", RPCs.ObjectHasRPCsQueuedOfType(ObjecRef.Entity, ReliableType));
+
+    return true;
+}
+
+TEST(FRPCContainer, GIVEN_a_container_WHEN_six_values_of_different_type_have_been_added_and_processed_THEN_they_have_been_processed_in_order)
+{
+	UMockObject* TargetObject = NewObject<UMockObject>();
+	FUnrealObjectRef ObjectRef = GenerateObjectRef(TargetObject);
+	FRPCContainer RPCs;
+
+	TMap<ESchemaComponentType, TArray<uint32>> RPCIndices;
+
+	for (int i = 0; i < 4; ++i)
+	{
+		FPendingRPCParamsPtr ParamsUnreliable = CreateMockParameters(TargetObject, UnreliableType);
+		FPendingRPCParamsPtr ParamsReliable = CreateMockParameters(TargetObject, ReliableType);
+		FPendingRPCParamsPtr ParamsCrossServer = CreateMockParameters(TargetObject, CrossServerType);
+		FPendingRPCParamsPtr ParamsMulticast = CreateMockParameters(TargetObject, MulticastType);
+
+		RPCIndices.FindOrAdd(UnreliableType).Push(ParamsUnreliable->Payload.Index);
+		RPCIndices.FindOrAdd(ReliableType).Push(ParamsReliable->Payload.Index);
+		RPCIndices.FindOrAdd(CrossServerType).Push(ParamsCrossServer->Payload.Index);
+		RPCIndices.FindOrAdd(MulticastType).Push(ParamsMulticast->Payload.Index);
+
+		RPCs.QueueRPC(MoveTemp(ParamsUnreliable), UnreliableType);
+		RPCs.QueueRPC(MoveTemp(ParamsReliable), ReliableType);
+		RPCs.QueueRPC(MoveTemp(ParamsCrossServer), CrossServerType);
+		RPCs.QueueRPC(MoveTemp(ParamsMulticast), MulticastType);
+	}
+
+	FProcessRPCDelegate Delegate;
+	Delegate.BindUObject(TargetObject, &UMockObject::ProcessRPC);
+	RPCs.ProcessRPCs(Delegate);
+
+	bool bProcessedInOrder = true;
+	for (const auto& processedIndicesOfType : TargetObject->ProcessedRPCIndices)
+	{
+		TArray<uint32>* storedIndicesOfType = RPCIndices.Find(processedIndicesOfType.Key);
+		if (!storedIndicesOfType)
+		{
+			bProcessedInOrder = false;
+			break;
+		}
+
+		if (processedIndicesOfType.Value.Num() != (*storedIndicesOfType).Num())
+		{
+			bProcessedInOrder = false;
+			break;
+		}
```

Comment:
These 2 conditions don't seem to be about order, more that each specific type of RPC is processed the right number of times. That would be a new test I think... if it's significant in contrast to the RPC instances each having been processed.

Reply:
These conditions check that the order per each RPC type is preserved. Is there something that should be changed to make the intention clearer?

Reply:
I might be misunderstanding this but the way I'm reading it the first condition is checking that the type gets processed at all. The second checks that each type gets processed the expected number of times. Only the 3rd condition checks that the call sequence is honoured, which is what the test says it is looking for. Therefore the 3rd condition could stand alone in this test (given the test name) and the other 2 conditions could then be either 1 or 2 separate tests, or perhaps not required at all depending on what the behaviour of the container should be. As it stands, when those 2 fail then it will not be an indicator that the order is wrong.

Reply:
These checks are there since failing any of those conditions will result in crash. However, I could change it to use check() instead of conditions.

Reply:
Using check() instead of conditions now.